src/pyupgrade/plugins/rewrite_yield_from.rs

```diff
@@ -0,0 +1,161 @@
+        match &stmt.node {
+            StmtKind::For {
+                target,
+                body,
+                orelse,
+                iter,
+                ..
+            } => {
+                // If there is an else statement, don't rewrite.
+                if !orelse.is_empty() {
+                    return;
+                }
+                // If there's any logic besides a yield, don't rewrite.
+                if body.len() != 1 {
+                    return;
+                }
+                // If the body is not a yield, don't rewrite.
+                let body = &body[0];
+                if let StmtKind::Expr { value } = &body.node {
+                    if let ExprKind::Yield { value: Some(value) } = &value.node {
+                        let names = collect_names(target);
+                        if names == collect_names(value) {
+                            self.yields.push(YieldFrom {
+                                stmt,
+                                body,
+                                iter,
+                                names,
+                            });
+                        }
+                    }
+                }
+            }
+            StmtKind::FunctionDef { .. }
+            | StmtKind::AsyncFunctionDef { .. }
+            | StmtKind::ClassDef { .. } => {
+                // Don't recurse into anything that defines a new scope.
+            }
+            _ => visitor::walk_stmt(self, stmt),
+        }
+    }
+
+    fn visit_expr(&mut self, expr: &'a Expr) {
+        match &expr.node {
+            ExprKind::ListComp { .. }
+            | ExprKind::SetComp { .. }
+            | ExprKind::DictComp { .. }
+            | ExprKind::GeneratorExp { .. }
+            | ExprKind::Lambda { .. } => {
+                // Don't recurse into anything that defines a new scope.
+            }
+            _ => visitor::walk_expr(self, expr),
+        }
+    }
+}
+
+#[derive(Default)]
+struct ReferenceVisitor<'a> {
+    parent: Option<&'a Stmt>,
+    references: FxHashMap<RefEquality<'a, Stmt>, Vec<&'a str>>,
+}
+
+impl<'a> Visitor<'a> for ReferenceVisitor<'a> {
+    fn visit_stmt(&mut self, stmt: &'a Stmt) {
+        let prev_parent = self.parent;
+        self.parent = Some(stmt);
+        visitor::walk_stmt(self, stmt);
+        self.parent = prev_parent;
+    }
+
+    fn visit_expr(&mut self, expr: &'a Expr) {
+        match &expr.node {
+            ExprKind::Name { id, ctx } => {
+                if matches!(ctx, ExprContext::Load | ExprContext::Del) {
+                    if let Some(parent) = self.parent {
+                        self.references
+                            .entry(RefEquality(parent))
+                            .or_default()
+                            .push(id);
+                    }
+                }
+            }
+            _ => visitor::walk_expr(self, expr),
+        }
+    }
+}
+
+/// UP028
+pub fn rewrite_yield_from(checker: &mut Checker, stmt: &Stmt) {
```

Comment:
@colin99d - I ended up refactoring this quite a bit. You might be interested in checking out the current implementation. The logic is roughly: do one pass over the function to find any candidate `yield` statements to convert, do a second pass to find any references to variables, then cross-reference to ensure we don't rewrite any yields whose variables are referenced elsewhere.